### PR TITLE
test(e2e): add autoscaling from zero e2e test

### PR DESCRIPTION
**Pull request by LiangquanLi930**
**What type of PR is this?**
/kind flake

**What this PR does / why we need it**:
Add end-to-end test that validates autoscaling-from-zero functionality using Cluster Autoscaler. Test creates MachineDeployment at 0 replicas, deploys Cluster Autoscaler, triggers scale-up with workload, and verifies machines scale correctly.
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] cherry-pick candidate

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by LiangquanLi930** [comment]
@jackfrancis @mboersma Hi, Could you help to add the ok-to-test label? Thanks!

File: test/e2e/kubernetes/deployment/deployment.go [hunk 1]
```diff
@@ -243,3 +243,15 @@ func (d *Builder) AddPodAntiAffinity(affinity corev1.PodAntiAffinity) *Builder {
 	d.deployment.Spec.Template.Spec.Affinity.PodAntiAffinity = &affinity
 	return d
 }
+
+// SetResourceRequests sets CPU and memory resource requests for the first container.
+// For multi-container pods, use SetResourceRequestsForContainer.
```
Comment (mboersma) [inline test/e2e/kubernetes/deployment/deployment.go]: Where is the `SetResourceRequestsForContainer` func this comment mentions? I couldn't find an implementation.

File: test/e2e/azure_autoscaling_from_zero.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+			"Type":   Equal(corev1.PodReady),
+			"Status": Equal(corev1.ConditionTrue),
+		})))
+	}, input.WaitIntervals...).Should(Succeed())
+
+	Byf("Cluster Autoscaler is running and ready in management cluster")
+
+	// Step 6: Create nginx deployment to trigger scale-up in workload cluster
+	By("Creating nginx deployment to trigger scale-up")
+	_, err = deploymentBuilder.Create("nginx:1.21", "autoscale-trigger", corev1.NamespaceDefault).
```
Comment (mboersma) [inline test/e2e/azure_autoscaling_from_zero.go]: This creates a Deployment but should probably also delete it after the test spec is done.

File: test/e2e/azure_autoscaling_from_zero.go [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+	// Add autoscaler annotations
+	if md.Annotations == nil {
+		md.Annotations = make(map[string]string)
+	}
+	md.Annotations["cluster.x-k8s.io/cluster-api-autoscaler-node-group-min-size"] = "0"
+	md.Annotations["cluster.x-k8s.io/cluster-api-autoscaler-node-group-max-size"] = "3"
+
+	// Update the MachineDeployment
+	err = mgmtClient.Update(ctx, md)
+	Expect(err).NotTo(HaveOccurred())
```
Comment (mboersma) [inline test/e2e/azure_autoscaling_from_zero.go]: This could fail if the MachineDeployment is being modified concurrently. Maybe we should use `Eventually` instead of `Expect`?

File: test/e2e/azure_autoscaling_from_zero.go [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+	// Step 7: Wait for CA to scale MachineDeployment from 0 to 1+
+	By("Waiting for Cluster Autoscaler to scale MachineDeployment to 1+")
+	Eventually(func(g Gomega) {
+		md := &clusterv1.MachineDeployment{}
+		err := mgmtClient.Get(ctx, client.ObjectKeyFromObject(autoscaleMD), md)
+		g.Expect(err).NotTo(HaveOccurred())
+		g.Expect(md.Spec.Replicas).NotTo(BeNil())
+		g.Expect(*md.Spec.Replicas).To(BeNumerically(">=", 1))
```
Comment (mboersma) [inline test/e2e/azure_autoscaling_from_zero.go:161]: Should we also verify that `Status.Replicas` and `Status.ReadyReplicas` reached the desired count? Maybe also that Nodes were actually added to the workload cluster?

File: test/e2e/azure_test.go [hunk 5]
```diff
@@ -1422,4 +1461,66 @@ var _ = Describe("Workload cluster creation", func() {
...
+
+				for _, name := range clusterRoles {
+					cr := &rbacv1.ClusterRole{}
+					cr.Name = name
+					_ = mgmtClient.Delete(ctx, cr)
+
+					crb := &rbacv1.ClusterRoleBinding{}
+					crb.Name = name
+					_ = mgmtClient.Delete(ctx, crb)
+				}
```
Comment (mboersma) [inline test/e2e/azure_test.go:1492]: We should also clean up the ServiceAccount we created in `rbac.yaml`.

File: test/e2e/azure_autoscaling_from_zero.go [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+type AutoscalingFromZeroSpecInput struct {
+	BootstrapClusterProxy framework.ClusterProxy
+	Namespace             *corev1.Namespace
+	ClusterName           string
+	Cluster               *clusterv1.Cluster
+	WaitIntervals         []interface{}
+}
+
+// AutoscalingFromZeroSpec implements a test that verifies autoscaling from zero functionality.
+// It creates a MachineDeployment scaled to 0, deploys Cluster Autoscaler, triggers scale-up
```
Comment (mboersma) [inline test/e2e/azure_autoscaling_from_zero.go]: Nitpick: this doesn't seem to *create* a MachineDeployment, just modify an existing one.

File: test/e2e/kubernetes/deployment/deployment.go (hunk 1 above)
Comment (LiangquanLi930) [reply test/e2e/kubernetes/deployment/deployment.go]: Sorry, this was my mistake. SetResourceRequests is sufficient for the current testing needs, so this comment should be removed.

File: test/e2e/azure_test.go (hunk 5 above)
Comment (LiangquanLi930) [reply test/e2e/azure_test.go:1492]: IIUC, ServiceAccount (and Deployment) are namespace-scoped resources that will be automatically cleaned up when the namespace is deleted, https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/38ff4480368453708593c6d0b7c72d0644727ddd/test/e2e/common.go#L200-L204

File: test/e2e/azure_autoscaling_from_zero.go (hunk 4 above)
Comment (LiangquanLi930) [reply test/e2e/azure_autoscaling_from_zero.go:161]: done

**Review by mboersma — APPROVED** [review]
/lgtm
/label tide/merge-method-squash
/assign @jackfrancis

**Comment by LiangquanLi930** [comment]
@mboersma, I’m working on fixing the pull-cluster-api-provider-azure-apiversion-upgrade job.

**Comment by mboersma** [comment]
@LiangquanLi930 it may be fixed by #6077.

**Comment by LiangquanLi930** [comment]
@jackfrancis When you have time, could you help to review this PR? Thanks!

**Review by mboersma — APPROVED** [review]
/approve
